utils/calendrical_calculations/src/islamic.rs

```diff
@@ -0,0 +1,271 @@
+use crate::astronomy::*;
+use crate::helpers::{self, div_rem_euclid, div_rem_euclid64, div_rem_euclid_f64, next};
+use crate::rata_die::{Moment, RataDie};
+
+// Different islamic calendars use different epochs (Thursday vs Friday) due to disagreement on the exact date of Mohammed's migration to Mecca.
+/// Lisp code reference: <https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L2066>
+const FIXED_ISLAMIC_EPOCH_FRIDAY: RataDie = crate::julian::fixed_from_julian(622, 7, 16);
+const FIXED_ISLAMIC_EPOCH_THURSDAY: RataDie = crate::julian::fixed_from_julian(622, 7, 15);
+
+/// Lisp code reference: <https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6898>
+const CAIRO: Location = Location {
+    latitude: 30.1,
+    longitude: 31.3,
+    elevation: 200.0,
+    zone: (1_f64 / 12_f64),
+};
+
+// "Fixed" is a day count representation of calendars staring from Jan 1st of year 1 of the Georgian Calendar.
+// The fixed date algorithms are from
+// Dershowitz, Nachum, and Edward M. Reingold. _Calendrical calculations_. Cambridge University Press, 2008.
+/// Lisp code reference: <https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6904>
+pub fn fixed_from_islamic_observational(year: i32, month: u8, day: u8) -> RataDie {
+    let year = i64::from(year);
+    let month = i64::from(month);
+    let day = i64::from(day);
+    let midmonth = FIXED_ISLAMIC_EPOCH_FRIDAY.to_f64_date()
+        + (((year - 1) as f64) * 12.0 + month as f64 - 0.5) * MEAN_SYNODIC_MONTH;
+    let lunar_phase =
+        Astronomical::calculate_lunar_phase_at_or_before(RataDie::new(midmonth as i64));
+    Astronomical::phasis_on_or_before(RataDie::new(midmonth as i64), CAIRO, Some(lunar_phase)) + day
+        - 1
+}
+
+/// Lisp code reference: <https://github.com/EdReingold/calendar-code2/blob/1ee51ecfaae6f856b0d7de3e36e9042100b4f424/calendar.l#L6983-L6995>
+pub fn observational_islamic_from_fixed(date: RataDie) -> (i32, u8, u8) {
```

Comment:
naming question: why is this function named `observational_islamic...`, but the function above is `...islamic_observational`?

Reply:
will clean up in a followup

Reply:
ah, this is from the book. I'm going to stick to that naming